server/service/src/invoice/inbound_shipment/update/generate.rs

```diff
@@ -398,34 +372,89 @@ pub fn generate_lines_and_stock_lines(
-        enter_datetime: Some(Utc::now().naive_utc()),
-        exit_datetime: None,
-    }
+    batch: &Vec<LineAndStockLine>,
+) -> Vec<LocationMovementRow> {
+    batch
+        .iter()
+        .filter_map(|batch| {
+            batch
+                .stock_line
+                .as_ref()
+                .map(|stock_line| LocationMovementRow {
+                    id: uuid(),
+                    store_id: store_id.clone(),
+                    stock_line_id: stock_line.id.clone(),
+                    location_id: batch.line.location_id.clone(),
+                    enter_datetime: Some(Utc::now().naive_utc()),
+                    exit_datetime: None,
+                })
+        })
+        .collect()
 }
 
-fn is_updating_donor(update_donor_method: Option<UpdateDonorMethod>) -> bool {
-    // should update batches if donor_id for invoice changes and change line donor_ids also selected
-    if let Some(donor_update_method) = update_donor_method {
-        // only update lines if a method of updating is selected.
-        // note allowing update if no donor_id is supplied for update to be conducted on existing donor_id
-        if donor_update_method != UpdateDonorMethod::NoChanges {
-            return true;
+fn update_donor_on_lines_and_stock(
+    connection: &StorageConnection,
+    invoice_id: &str,
+    updated_default_donor_id: Option<String>,
+    donor_update_method: Option<UpdateDonorLineMethod>,
+) -> Result<Vec<LineAndStockLine>, UpdateInboundShipmentError> {
+    let invoice_lines = InvoiceLineRepository::new(connection).query_by_filter(
+        InvoiceLineFilter::new()
+            .invoice_id(EqualFilter::equal_to(invoice_id))
+            .r#type(InvoiceLineType::StockIn.equal_to()),
+    )?;
+    let mut result = Vec::new();
+
+    for invoice_line in invoice_lines {
+        let mut line = invoice_line.invoice_line_row;
+        let mut stock_line = invoice_line.stock_line_option;
+
+        let new_donor_id = match donor_update_method.clone() {
+            Some(UpdateDonorLineMethod::NoChanges) | None => line.donor_id.clone(),
+            Some(UpdateDonorLineMethod::UpdateExistingDonor) => match line.donor_id {
+                Some(_) => updated_default_donor_id.clone(),
+                None => None,
+            },
+            Some(UpdateDonorLineMethod::AssignIfNone) => {
+                line.donor_id.clone().or(updated_default_donor_id.clone())
+            }
+            Some(UpdateDonorLineMethod::AssignToAll) => updated_default_donor_id.clone(),
+        };
+
+        line.donor_id = new_donor_id.clone();
+        if let Some(ref mut stock_line) = stock_line {
+            stock_line.donor_id = new_donor_id;
         }
+
+        result.push(LineAndStockLine { line, stock_line });
     }
-    return false;
+
+    Ok(result)
+}
+
+fn should_update_donor(
+    donor_id: Option<String>,
+    update_donor_method: &Option<UpdateDonorLineMethod>,
+) -> bool {
+    if donor_id.is_none() {
```

Comment:
We want them to always state a reason so have made an early return here... but hmm idk 🤔

Reply:
what do you mean a reason?

Reply:
UpdateDonorLineMethod <- we pop this up if they change default_donor_id, so enforcing reason selection